lib/task_helpers/imports/tags.rb

```diff
@@ -0,0 +1,88 @@
+module TaskHelpers
+  class Imports
+    class Tags
+      class ClassificationYamlError < StandardError
+        attr_accessor :details
+
+        def initialize(message = nil, details = nil)
+          super(message)
+          self.details = details
+        end
+      end
+
+      def import(options = {})
+        return unless options[:source]
+
+        glob = File.file?(options[:source]) ? options[:source] : "#{options[:source]}/*.yaml"
+        Dir.glob(glob) do |fname|
+          begin
+            tag_categories = YAML.load_file(fname)
+            import_tags(tag_categories)
+          rescue ClassificationYamlError => e
+            warn("Error importing #{fname} : #{e.message}")
+            e.details.each { |d| warn("\t#{d}") }
+          rescue ActiveModel::UnknownAttributeError => e
+            warn("Error importing #{fname} : #{e.message}")
+          end
+        end
+      end
+
+      private
+
+      # Tag Categories that are not visible in the UI and should not be imported
+      SPECIAL_TAGS = %w(/managed/folder_path_yellow /managed/folder_path_blue /managed/user/role).freeze
+
+      UPDATE_FIELDS = %w(description example_text show perf_by_tag).freeze
+
+      REGION_NUMBER = MiqRegion.my_region_number.freeze
+
+      def import_tags(tag_categories)
+        tag_categories.each do |tag_category|
+          tag = tag_category["ns"] ? "#{tag_category["ns"]}/#{tag_category["name"]}" : "/managed/#{tag_category["name"]}"
+          next if SPECIAL_TAGS.include?(tag)
+          Classification.transaction do
+            import_classification(tag_category)
+          end
+        end
+      end
+
+      def import_classification(tag_category)
+        ns = tag_category["ns"] ? tag_category["ns"] : "/managed"
+        tag_category["name"] = tag_category["name"].to_s
+
+        classification = Classification.find_by_name(tag_category['name'], REGION_NUMBER, ns, 0)
+
+        entries = tag_category.delete('entries')
+
+        if classification
+          classification.update_attributes!(tag_category.select { |k| UPDATE_FIELDS.include?(k) })
+        else
+          classification = Classification.create(tag_category)
+          raise ClassificationYamlError.new("Tag Category error", classification.errors.full_messages) unless classification.valid?
```

Comment:
I'm not sure I'm completely following. Are you saying to call `Classification.new` regardless of if the Classification already exists or not?

I've removed all of the additional checks I was doing and letting the model handle that. I do have to make sure that the `name` attribute for a Classification or Entry is a string. If it is `nil` from reading in the YAML file then the `find_by_name` and `find_entry_by_name` (lines 53 & 71) calls cause the rake task to be aborted with a `TypeError: no implicit conversion of nil into String`. So by making sure `name` is a string I can just depend on the models validations.

I see what you mean with the `update_attributes` so I'll make a change to handle errors when updating the attributes.

I left the one error class `ClassificationYamlError` so that I could output a bit nicer format of the validation errors for the user.